pkg/render/common/components/components_test.go

```diff
@@ -555,8 +556,522 @@ var _ = Describe("Common components render tests", func() {
+					Spec: &v1.TyphaDeploymentSpec{
+						Template: &v1.TyphaDeploymentPodTemplateSpec{
+							Spec: &v1.TyphaDeploymentPodSpec{
+								NodeSelector: map[string]string{},
+							},
+						},
+					},
+				}
+			},
+			func(result appsv1.Deployment) {
+				expected := defaultedDeployment()
+				Expect(result.Spec.Template.Spec.NodeSelector).To(Equal(expected.Spec.Template.Spec.NodeSelector))
+				Expect(result).To(Equal(expected))
+			}),
+
+		Entry("affinity, nodeSelector, and tolerations",
+			defaultedDeployment,
+			func() *v1.TyphaDeployment {
+				return &v1.TyphaDeployment{
+					Spec: &v1.TyphaDeploymentSpec{
+						Template: &v1.TyphaDeploymentPodTemplateSpec{
+							Spec: &v1.TyphaDeploymentPodSpec{
+								Affinity:     &affinity,
+								NodeSelector: nodeSelector,
+								Tolerations:  tolerations,
+							},
+						},
+					},
+				}
+			},
+			func(result appsv1.Deployment) {
+				expected := defaultedDeployment()
+				expected.Spec.Template.Spec.Affinity = &affinity
+				// only keys that don't already exist in the nodeSelector are added
+				for k, v := range nodeSelector {
+					if _, ok := expected.Spec.Template.Spec.NodeSelector[k]; !ok {
+						expected.Spec.Template.Spec.NodeSelector[k] = v
+					}
+				}
+				expected.Spec.Template.Spec.Tolerations = tolerations
+				Expect(result.Spec.Template.Spec.Affinity).To(Equal(expected.Spec.Template.Spec.Affinity))
+				Expect(result.Spec.Template.Spec.NodeSelector).To(Equal(expected.Spec.Template.Spec.NodeSelector))
+				Expect(result.Spec.Template.Spec.Tolerations).To(Equal(expected.Spec.Template.Spec.Tolerations))
+				Expect(result).To(Equal(expected))
+			}),
+
+		Entry("terminationGracePeriod",
+			defaultedDeployment,
+			func() *v1.TyphaDeployment {
+				return &v1.TyphaDeployment{
+					Spec: &v1.TyphaDeploymentSpec{
+						Template: &v1.TyphaDeploymentPodTemplateSpec{
+							Spec: &v1.TyphaDeploymentPodSpec{
+								TerminationGracePeriodSeconds: int64Ptr(3),
+							},
+						},
+					},
+				}
+			},
+			func(result appsv1.Deployment) {
+				Expect(*result.Spec.Template.Spec.TerminationGracePeriodSeconds).To(Equal(int64(3)))
+			}),
+
+		Entry("strategy",
+			defaultedDeployment,
+			func() *v1.TyphaDeployment {
+				return &v1.TyphaDeployment{
+					Spec: &v1.TyphaDeploymentSpec{
+						Strategy: &appsv1.DeploymentStrategy{
+							Type: appsv1.RollingUpdateDeploymentStrategyType,
+							RollingUpdate: &appsv1.RollingUpdateDeployment{
+								MaxUnavailable: intOrStrPtr("0"),
+								MaxSurge:       intOrStrPtr("100%"),
+							},
+						},
+					},
+				}
+			},
+			func(result appsv1.Deployment) {
+				Expect(result.Spec.Strategy).To(Equal(appsv1.DeploymentStrategy{
+					Type: appsv1.RollingUpdateDeploymentStrategyType,
+					RollingUpdate: &appsv1.RollingUpdateDeployment{
+						MaxUnavailable: intOrStrPtr("0"),
+						MaxSurge:       intOrStrPtr("100%"),
+					},
+				}))
+			}),
+	)
 })
 
+func intOrStrPtr(v string) *intstr.IntOrString {
```

Comment:
We have a package in the operator for these sorts of functions already, so might be good to put this one there and use the existing int64 function: https://github.com/tigera/operator/blob/master/pkg/ptr/conversion.go